### ISSUE TITLE
Writing a sufficiently large string to a git bash pipe on windows regressed in 1.62

### ISSUE THREAD

**Issue by smmalis37** [rust-lang/rust#98947]
### Reproduction steps:
1. Be on windows.
2. `cargo new --bin something`
3. replace main.rs with [main.txt](https://github.com/rust-lang/rust/files/9048190/main.txt) and rename it back to main.rs.
4. Open git bash
5. Run `cargo run | cat`

I expected to see this happen: my really long string gets printed twice, followed by a clean exit

Instead, this happened: some amount of my really long string gets printed, followed by: 
`fatal runtime error: I/O error: operation failed to complete synchronously`

### Additional Details
- The test case, while simple, does appear to have a minimum size requirement to trigger this crash. Shrinking the string will cause it to stop crashing.
- I'm using `cat` in the steps for simplicity, but we've observed this happening with other programs at the end of the pipe too.
- Redirecting to a file instead of a pipe seems to not trigger this crash.
- Using `cargo run` is not necessary to trigger the crash, running the built executable directly also crashes.
- This PR added this new error condition: https://github.com/rust-lang/rust/pull/95469
- I am not the only one seeing this new error message: https://github.com/rust-lang/rust/issues/98700#issuecomment-1171654196

### Version it worked on
```
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-pc-windows-msvc
release: 1.61.0
LLVM version: 14.0.0
```

### Version with regression
```
rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-pc-windows-msvc
release: 1.62.0
LLVM version: 14.0.5
```

@rustbot modify labels: +regression-from-stable-to-stable +O-windows

**Comment by jstarks** [comment]
@ChrisDenton, I believe this must be because git bash (cygwin? msys?) is opening pipes between processes with `FILE_FLAG_OVERLAPPED`.

In my original bug report, I suggested calling `GetOverlappedResult` once if `NtReadFile` returned `STATUS_PENDING`, since that will at least resolve the pending IO correctly in the case where the file is not being used for other IOs concurrently. This would cover this and other cases at least as well as the original code calling `ReadFile`. This isn't perfect, since of course `File::read` takes `&self`, so you can probably deadlock in a multi-threaded process in this async pipe configuration. But it would at least avoid the common case.

However, it seems that this got missed from the actual change (sorry, I should have caught that). Is there a reason this was skipped?

**Comment by t1m0thyj** [comment]
Thanks for fixing the issue quickly. 😀 When the PR that fixes this issue is merged, will a patch release 1.62.1 be published? 🙏 Or will we have to wait for a fix in 1.63?
